src/tools/cephfs_mirror/PeerReplayer.cc

```diff
@@ -826,8 +977,10 @@ int PeerReplayer::remote_file_op(const std::string &dir_root, const std::string
   return 0;
 }
 
-int PeerReplayer::cleanup_remote_dir(const std::string &dir_root,
-                                     const std::string &epath, const FHandles &fh) {
+int PeerReplayer::cleanup_remote_dir(
```

Comment:
Potentially cleanup_remote_dir & propagate_deleted_entries could benefit from multi-threaded  processing as well. One might e.g. collect all the to-be-processed entries at specific level and form their processing as a task or make a task from a subfolder or something. There might be some performance benefit on large directory trees as well.
Perhaps could be implemented as a second stage, not with this patch?

Reply:
I would expect metadata-only operations to be fast enough anyway, and exclude the proposed optimization from the first iteration of the patch. The big question here is: what do we do with directories containing millions of deleted subdirs? Wouldn't collecting them in advance balloon the memory usage? Can we queue such mass deletions piece-wise without storing the whole list of stuff to delete?

Reply:
For deletion, I agree it is not possible without keeping a list in memory but for creation of dirs we can do that as, while creation we are going from top to down.

Reply:
Also in the current implementation of `main` branch it is doing bfs for syncing snapshot having id >= 2. So it is actually consuming memory of at least maximum number of nodes in the same level.